File: launcher/src/org/koreader/device/rockchip/RK33xxEPDController.java
/* generic EPD Controller for Android devices
 *
 * interface for boyue mimas and maybe others,
 * based on https://github.com/koreader/koreader/issues/4595 */

package org.koreader.device.rockchip;

import android.util.Log;


@SuppressWarnings("unchecked")
public abstract class RK33xxEPDController {
    private static final String TAG = "epd";
    public static final int EPD_FULL = 1;
    public static final int EPD_A2 = 2;
    public static final int EPD_PART = 3;
    public static final int EPD_BLACK_WHITE = 6;
    public static final int EPD_FORCE_FULL = 11;
    public static final int EPD_REGAL = 15;
    public static final int EPD_ADAPTATIVE = 17;
    public static final int EPD_FAST = 18;
    public static final int EPD_DITHER = 100;

    public static boolean requestEpdMode(String epdMode) {
        try {
            Class.forName("android.view.View").getMethod("setByEinkUpdateMode",
                new Class[] { Integer.TYPE }).invoke(null, new Object[] { getEpdMode(epdMode) });
            return true;
        } catch (Exception e) {
            Log.e(TAG, e.toString());
            return false;
        }
    }

    private static int getEpdMode(String epdMode) {
        /** only full mode was tested, so return that.
         *  device owners will want to expand this
         *  to return other well known modes
         */
        int mode = EPD_FULL;
        Log.v(TAG, String.format("Requesting %s: %d", epdMode, mode));
        return mode;
    }
}

File: app/src/main/java/org/koreader/launcher/device/epd/rockchip/RK33xxEPDController.kt
/* generic EPD Controller for Android devices
 *
 * interface for boyue rk33xx clones, like the Likebook Mimas, Mars, Muses...
 * based on https://github.com/koreader/koreader/issues/4595
 *
 * Thanks to @carlinux
 *
 * Only full mode was tested, so 'getEpdMode' returns that.
 * Device owners will want to expand this to return other well known modes:
 *
 * val EPD_A2 = 2
 * val EPD_PART = 3
 * val EPD_BLACK_WHITE = 6
 * val EPD_FORCE_FULL = 11
 * val EPD_REGAL = 15
 * val EPD_ADAPTATIVE = 17
 * val EPD_FAST = 18
 * val EPD_DITHER = 100
 */

package org.koreader.launcher.device.epd.rockchip

import android.util.Log
import java.util.*

abstract class RK33xxEPDController {
    companion object {
        private const val TAG = "EPD"
        private const val EPD_FULL = 1

        fun requestEpdMode(epdMode: String): Boolean {
            return try {
                Class.forName("android.view.View").getMethod("setByEinkUpdateMode",
                        Integer.TYPE).invoke(null, getEpdMode(epdMode))
                true
            } catch (e: Exception) {
                Log.e(TAG, e.toString())
                false
            }
        }

        private fun getEpdMode(epdMode: String): Int {
            val mode = EPD_FULL
            Log.v(TAG, String.format(Locale.US, "Requesting %s: %d", epdMode, mode))
            return mode
        }
    }
}
